### DISCUSSION TITLE
3D Tiles

### DISCUSSION THREAD

**Discussion by donnyv** [maplibre/maplibre-gl-js#3378] [Ideas]
With Google changing their policy and letting anyone include their Google Earth buildings as GlTF. Maybe Maplibre can add it as a native layer source.  
https://cloud.google.com/blog/products/maps-platform/create-immersive-3d-map-experiences-photorealistic-3d-tiles

**Comment by HarelM** [comment]
Can you clarify the issue? Problem statement? Feature request?
Do you mean that maplibre should support 3D tiles as part of this issue?

**Comment by IvanSanchez** [comment]
Your premise ("Google let's[sic] anyone use Google Earth buildings as 3D Tiles") is wrong. Let me quote the linked press release, emphasis mine:

> Today we're launching the Experimental release of Photorealistic 3D Tiles, **available through the Map Tiles API**.

Now let me quote their [3D tiles documentation](https://developers.google.com/maps/documentation/tile/3d-tiles):

> Before you start using the Photorealistic 3D Tiles API, you need a project with a billing account, and you need the Map Tiles API enabled.

That is not "letting anyone", and there has been no policy change. Please stop the spread of openwashing.

**Comment by donnyv** [comment]
@IvanSanchez Sorry I didn't mean too to spread "openwashing". Which until now I didn't know that word existed LOL I didn't realize the data wasn't actually open sourced. Just free to use there service....for now.

As for support for 3D Tiles. I still think it would be beneficial to have it in maplibra.  
https://github.com/CesiumGS/3d-tiles

**Comment by prozessor13** [comment]
Beside google, other cities, like vienna, offers 3dtiles (https://www.wien.gv.at/stadtplan3d/ click on Content and then activate 3D-Mesh). So I think this would be really a great addition to maplibre, too.
But implementing 3dtiles/gltf into maplibre will be, for sure, a big project. May it is better to render it with deck.gl, beside with maplibre. Deck.gl for 3dTiles, maplibre for labels and markers.

Would be great to start a discussion, especially some thoughts by people who know deck.gl, how a potential implementation could look like.

**Comment by wipfli** [comment]
I wonder what the best approach is for 3D content in MapLibre. Already now, you can integrate single glTF models in MapLibre GL JS with Three.JS or deck.gl.

But that is just single 3D objects with a fixed level of detail, a bit like Apple maps does it in the vector version or also Snap if I am not wrong.

**Comment by IvanSanchez** [comment]
The 3D tiles spec is a bit of a monster - see https://github.com/CesiumGS/3d-tiles/tree/main/specification .

TL;DR: Tiles can have one of several shapes (flat squares; subes; spheres) and can have one of several data formats (pointclouds, glTF). They behave like R-trees, rather than like tiles.

**Comment by prozessor13** [comment]
I agree that 3DTiles are a bit of a monster. I had a requirement to render CityGML Data onto maplibre, and i ended up reusing the VectorTile spezification with a custom geometry-type with 3 coordinates instead of 2d. This had the advantage the all the tile-covering logic of maplibre can be reused. But in the end it is a custom solution which is not future-proof. So, beacuse now "everyone" is using 3DTiles/gltf I think there is no way around. I also do not know some kind of lightweight library to load/parse 3DTiles, and because deck.gl has already an implementation, why not use it?

If you want to have a look onto my custom-3d-vectortiles CityGML Demopage:
   https://static.maptoolkit.net/maplibre/buildings.html#15.23/48.210618/16.372333/-20/74
Except the lack of render textures onto the 3d objects, it is really performant, and was not very much work implementing/integrating it into maplibre.

**Reply by radityagumay** [reply]
hi @prozessor13 would you mind to share the link again? i got 404. thank you

**Reply by prozessor13** [reply]
hi, the 3d-buildings are used in the new vienna map. https://stadtplan.wien.gv.at/

**Comment by wipfli** [comment]
Looks amazing, thanks for sharing, @prozessor13 !

**Comment by StylianosVitalis-TomTom** [comment]
@prozessor13 This is awesome! It performs impressively, as well!

If I get it correctly, then you had to implement your own exporter and MapLibe GL JS parser for this customized 3D vector tiles?

**Comment by prozessor13** [comment]
Yes, i made a custom vector-tile server which adds to the existing 2d point/line/polygon vectortile-geometries a custom geometry with 3d polygons. And a custom layer (named fill-3d) in maplibre, which renders these geometries. Vertex normals are calculated in maplibre.

**Comment by StylianosVitalis-TomTom** [comment]
Would you be interested in sharing this solution? I think that this is an interesting direction for incorporating (trully) 3D data in MapLibre GL JS.

**Comment by prozessor13** [comment]
Sorry, this is a product of our company :) I only wanted to demonstrate a way beside 3DTiles, and for sure this format could be sufficient to render 3d-meshes (may even more performant and memory-friendly), and i really would like to implement this, but because everyone else uses 3DTiles, i think this is not the future.

**Comment by StylianosVitalis-TomTom** [comment]
Absolutely! Still, impressive work. Top notch! 🙌

I remain of the opinion that the whole 3D Tiles specification is an overkill (for non-mesh data, at least). It fits the Google use case, but for data such as in your example I think there are elements of 3D Tiles we can reuse (like the glTF+extension to hold attributes), but the tile definition specification is adding too much complexity, both to the implementation and the data processing part.

**Comment by wipfli** [comment]
If someone wants to implement 3d tiles in MapLibre, then in principle there is some funding available from cesium via the ecosystem grants program: https://cesium.com/cesium-ecosystem-grants/

**Comment by jsehner** [comment]
LoD2 buildings as 3D Tiles directly in MapLibre would be great! This addon works well, which was developed for Mapbox GL JS, but also works fine with MapLibre: https://github.com/Geodan/mapbox-3dtiles. It was use for the visualisation of Germany for basemap.de. A beta-service is released for Cesium (buildings on terrain) and MapLibre (buildings on sea level). Simple Viewer: https://web3d.basemap.de/maplibre/lod2.html#16.3/50.941485/6.959092/-39.2/57. Description: https://basemap.de/beta-lod2/. What is missing: It doesn't work in combination with the 3D Terrain, only with flat terrain. By the way: A different colour for the roofs is important in my opinion.

**Comment by gberaudo** [comment]
A few remarks:
- 3d-tiles is an OGC standard;
- [3d-tiles next](https://github.com/CesiumGS/3d-tiles/tree/main/next) is the next iteraction of the standard, which will include some simplifications (direct use of gltf, implicit tiling scheme, ...).

**Comment by StylianosVitalis-TomTom** [comment]
As far as I understand, 3D Tiles next is already part of the official standard, since version 1.1 which was approved and published a few months ago. Old formats such as b3dm etc. are considered already legacy. See here: https://docs.ogc.org/cs/22-025r4/22-025r4.html#tileformats-gltf-migration-from-legacy-tile-formats

**Comment by beifeizhou** [comment]
> 

I have been experimented mapbox with 3d tiles, the example Geodan repo is working well. However, it is not straightforward how the google 3d tile can be integrated, as it requires b3dm formats of the tiles.

**Comment by laem** [comment]
Hi, very interesting discussion. 

I was wondering if there is somewhere, a set of open source simplified 3d models of well-known buildings being built. It looks like MapBox [provides this](https://www.mapbox.com/blog/standard-core-style), [Apple Maps too](https://beta.maps.apple.com/?ll=48.859957512843685%2C2.327242583799631&spn=0.0025626138224339456%2C0.006600196929497315).

**Reply by HarelM** [reply]
I'm not sure the linked articles talk about 3D tiles, but more about 3D models, which are different... But I might be wrong.

**Reply by laem** [reply]
Yes, it's a different matter, they could be integrated with threejs

**Comment by jo-chemla** [comment]
Just a quick post to cross-reference existing issues in the maplibre-gl-js repo tied to this discussion: 
 - Adding 3D Tiles as a source for terrain https://github.com/maplibre/maplibre-gl-js/issues/4493 : despite 3D-tiles being full 3D and not 2.5D, it might be useful to use them as a source for 3D (terrain/buildings), hillshade, hypso etc. 
 - batched mesh support https://github.com/maplibre/maplibre-gl-js/issues/3794

I'm not sure whether the integration should happen within the maplibre ecosystem directly, or via deckgl [3DTiles](https://deck.gl/docs/api-reference/geo-layers/tile-3d-layer) layer.

**Comment by hh-hang** [comment]
I'd like to recommend an approach I'm currently using: MapLibre + Three.js + 3DTilesRendererJS. By integrating Three.js into MapLibre and using 3DTilesRendererJS to load 3D Tiles, it works quite well so far.
<img width="761" height="650" alt="1" src="https://github.com/user-attachments/assets/dfade091-f1cd-4fd7-bb84-52da5ec0a406" />

**Reply by HarelM** [reply]
If this can be made to an example as oart of the docs that can be great.

**Reply by hh-hang** [reply]
Sure, I can do that. I’ll prepare an example when I have time. However, I haven’t tried adding examples to the documentation before, so I may need some time to learn how to do it. Could you give me some guidance or suggestions?

**Reply by HarelM** [reply]
Yes, it's fairly simple, those are html pages in the example folder under tests. There's already a three js example you can copy and modify. After that you'd need to generate an image using a script, you should be able to find the relevant information in the readmes scattered in the repo. Let me know if you need more help.

**Reply by hh-hang** [reply]
Understood, thanks.

**Reply by hh-hang** [reply]
I have submitted the example for loading 3D Tiles. This is my first contribution. If there are any issues, please feel free to let me know. @HarelM 
https://github.com/maplibre/maplibre-gl-js/pull/7198

**Reply by safwat-halaby** [reply]
I'd like to improve your example. I added commentary in #7198

**Comment by safwat-halaby** [comment]
It is worth noting that #7523 makes room for much flexibility in 3d tiles. A transparent RGB terrain with `disableTerrainSkirts: true` allows us to create style spec polygons that trace the ground of a 3dtile model. Would  such an example be useful in the docs?
